Improve performance of rispy's parser and refactor parser

J535D165 opened the pull request:
This PR proposes to refactor and improve performance. Although breaking changes are not necessary, I would recommend changing parts of the API as well. Happy to hear your thoughts

## Performance 

I used PR #65 to test the performance of rispy on my machine.

Current main branch: **Time: 50.5917 s**
This PR: **Time: 20.3249 s**

According to this benchmark, this implementation is 2.5x faster. This seems to hold for both small and larger datasets. 

## No more regexp

This PR removes the need for regular expressions. Everything between the start and end tags is a tag with a specific format. So, you don't need to apply a regular expression to each line. This saves a lot of time and potential mistakes. 

I left the current regexp class attribute definition, but I would suggest to remove it.

## Issues resolved

This PR also resolves:

- #62 and #63 by @holub008. No extra changes were needed to make this work on the test defined by @holub008 in PR #63. 
- #61 by @holub008. Resolved by design. 

## API changes

We can simplify, improve, and refactor a bit more. I'll leave that for a moment until the community has looked over this proposal. 

## Other

I added PR #64's changes to this PR for convenience. They can/will be changed after a decision has been made on PR #64.

PeterLombaers commented on rispy/parser.py:
```diff
@@ -120,182 +104,124 @@ def __init__(
-                    self.in_ref = False
-                    self.last_tag = None
-                except NextLine:
+                if tag == "  ":
```
This assumes that all multiline content will be nicely formatted with sufficient whitespace so that the content starts at the same character each line. An example where this does not hold is [this test](https://github.com/MrTango/rispy/blob/main/tests/data/example_utf_chars.ris). It might be true for RIS files found in the wild, though?

holub008 replied:
+1, I run into this condition frequently (see the `example_empty_tag.ris` file I partially copied over from an Embase user).

The [prior implementation](https://github.com/MrTango/rispy/blob/d719b24023d136bcf8aa46ae34a90b3829e02a5c/rispy/parser.py#L164) supported this nicely, and a fix here may also address Peter's ^ comment.

J535D165 commented on rispy/parser.py:
```diff
@@ -120,182 +104,124 @@ def __init__(
-            if not line.strip():
-                continue
+            while True:
+                tag, content = self._parse_tag(next(lines))
```
I see. I changed the API now a bit. The naive (current, main-branch) implementation of methods `get_tag` and `get_content` are now replaced by `parse_line`. `parse_line` is now a lot stricter but not using slow regexp. It assumes a line between the start and end tag as described in the docstring. 

Does this fully address your feedback?

J535D165 commented:
The changes I made for the multiline parsing impacted the API. I chose to deprecate `get_tag` and `get_content` because there where zero guarantees that they actually return the tag and the content (the just slice the line). 

I took this freedom to break some other parts of the API and simplify things. I removed the regexp and deprecated BaseParser in favor of RisParser (no need to offer multiple options for the same thing). 

I'm happy to discuss these API changes with you. Would this impact you as a user, and how?

J535D165 commented:
I'm planning to make a PubMed parser (subclass of RisParser). I will use this as an experiment to see if this PR's API design is flexible enough to implement such a format (example https://github.com/asreview/citation-file-formatting/tree/main/Datasets/RIS).

PeterLombaers commented on rispy/parser.py:
```diff
@@ -120,226 +95,175 @@ def __init__(
+        tuple
+            Tuple containing the tag and the content of the tag.
+        """
+        if line[3:5] == " -" and line[:2].isupper():
```
This looks good, though you might as well test `line[2:5] == "  -"` here, right? That should have a lower chance of false positives for multiline content.

Basically this line is where all the parsing magic happen, and you can do it with string slicing or with regex. I ran a small test to see how these two compare:

```[python]
import regex

pattern = regex.compile(r"^[A-Z][A-Z0-9]  -")


def parse_with_regex(s):
    if bool(pattern.match(s)):
        return s[:2]


def parse_with_slice(s):
    if s[2:5] == "  -" and s[:2].isupper():
        return s[:2]


def test_parse_with_regex_match(benchmark):
    s = "AB  - this is a random text"
    assert benchmark(parse_with_regex, s) == "AB"


def test_parse_with_regex_no_match(benchmark):
    s = "this is a random text"
    assert benchmark(parse_with_regex, s) is None


def test_parse_with_slice_match(benchmark):
    s = "AB  - this is a random text"
    assert benchmark(parse_with_slice, s) == "AB"


def test_parse_with_slice_no_match(benchmark):
    s = "this is a random text"
    assert benchmark(parse_with_slice, s) is None
```

This gave the following results:
```
----------------------------------------------------------------------------------------------- benchmark: 4 tests ----------------------------------------------------------------------------------------------
Name (time in ns)                       Min                    Max                Mean              StdDev              Median                IQR            Outliers  OPS (Mops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_parse_with_slice_no_match     124.0000 (1.0)       1,167.0000 (1.0)      138.3356 (1.0)       31.6514 (1.0)      132.0000 (1.0)       7.0000 (1.33)    1885;2031        7.2288 (1.0)       76924         100
test_parse_with_slice_match        257.8947 (2.08)      7,605.2631 (6.52)     291.6303 (2.11)     101.2712 (3.20)     284.2105 (2.15)      5.2632 (1.0)    2315;22931        3.4290 (0.47)     196079          19
test_parse_with_regex_no_match     314.9999 (2.54)      6,020.0000 (5.16)     356.7073 (2.58)     119.1388 (3.76)     345.0000 (2.61)     15.0000 (2.85)    2646;3596        2.8034 (0.39)     149254          20
test_parse_with_regex_match        399.9994 (3.23)     28,300.0008 (24.25)    542.0298 (3.92)     359.8332 (11.37)    500.0002 (3.79)     99.9989 (19.00)       66;80        1.8449 (0.26)      34723           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

So string slicing seems the clear winner!

PeterLombaers replied:
Actually, the logic still isn't correct. This implementation will accept `1A  -` as a valid tag even though a digit is not allowed as the first character. You'd need something like `line[2:5] == "  -" and line[:2].isupper() and not line[:1].isdigit()`

J535D165 replied:
Resolved

PeterLombaers commented on rispy/parser.py:
```diff
@@ -120,226 +95,175 @@ def __init__(
+        if line[3:5] == " -" and line[:2].isupper():
+            return (line[0:2], line[6:].strip())
+        else:
+            return ("  ", line.strip())
```
Maybe return `None` instead of `"  "` for the empty tag? Then the boolean value of the empty tag will be false.

J535D165 replied:
Implemented. Thanks!

J535D165 commented:
Implemented suggestions and checks by @PeterLombaers. 

Updated benchmark results:

```
--------------------------------------------------------------------------------------------- benchmark: 2 tests --------------------------------------------------------------------------------------------
Name (time in ms)                             Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_rispy_large               241.2698 (1.0)      294.8977 (1.0)      259.4689 (1.0)      21.0500 (1.0)      252.0566 (1.0)      23.0623 (1.0)           1;0  3.8540 (1.0)           5           1
test_benchmark_rispy_large_multiline     282.0263 (1.17)     345.4905 (1.17)     300.7980 (1.16)     26.1683 (1.24)     288.5437 (1.14)     28.6715 (1.24)          1;0  3.3245 (0.86)          5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

Compared to main branch: 
```
--------------------------------------------------------------------------------------------- benchmark: 2 tests --------------------------------------------------------------------------------------------
Name (time in ms)                             Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_rispy_large               484.5452 (1.0)      517.4370 (1.0)      502.2159 (1.0)      14.4051 (1.0)      499.9848 (1.0)      25.6566 (1.0)           2;0  1.9912 (1.0)           5           1
test_benchmark_rispy_large_multiline     569.1948 (1.17)     626.2228 (1.21)     592.6684 (1.18)     27.2004 (1.89)     581.3098 (1.16)     50.2680 (1.96)          1;0  1.6873 (0.85)          5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

J535D165 commented:
Tested against this repo with many example exports: https://github.com/asreview/citation-file-formatting/tree/ecddcceff24791f5454f53f64515d5d0990649aa/Datasets/RIS. This PR will add support for Cochrane RIS exports (#61). All other example RIS files work except from outlier Pubmed.

shapiromatron commented on rispy/parser.py:
```diff
@@ -120,226 +95,175 @@ def __init__(
+        else:
+            return (None, line.strip())
+
+    def _add_single_value(self, record, name, value, is_multi=False):
```
add type annotations? at least `is_multi:bool = False`, are the others all str?

shapiromatron commented on rispy/parser.py:
```diff
@@ -120,226 +95,175 @@ def __init__(
+        except StopIteration:
+            return result
+
+    def parse_line(self, line):
```
could we easily add type annotations, are they just str?

shapiromatron reviewed (approved):
Looks great, a complete overhaul like you said but I think the code is more readable simpler. I had a few suggestions about adding type annotations if it's easy to do; I think they're largely just `str` so hopefully it's not too bad to add. 

Excited to get this one in finally!

shapiromatron commented:
@J535D165 Please let me know when you're ready to merge it in and I'll go ahead and do so.

J535D165 commented:
@shapiromatron Thanks for the feedback, all checks are green now. Feel free to merge.

J535D165 commented:
Received a real world example RIS file today of 650MB:

```
-------------------------------------------------------- benchmark: 1 tests -------------------------------------------------------
Name (time in s)                              Min      Max     Mean  StdDev   Median      IQR  Outliers     OPS  Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------
test_benchmark_rispy_big_example_file     16.7548  32.5059  25.0655  6.7390  27.9454  11.2026       2;0  0.0399       5           1
-----------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean

```
old:
```
------------------------------------------------------- benchmark: 1 tests -------------------------------------------------------
Name (time in s)                              Min      Max     Mean  StdDev   Median     IQR  Outliers     OPS  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------
test_benchmark_rispy_big_example_file     30.5232  43.5556  35.9390  5.2012  35.9779  7.8023       2;0  0.0278       5           1
----------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

I hope we can speed this up further in the near future (for example, with Rust).

shapiromatron commented:
> Received a real world example RIS file today of 650MB:
> 
> ```
> -------------------------------------------------------- benchmark: 1 tests -------------------------------------------------------
> Name (time in s)                              Min      Max     Mean  StdDev   Median      IQR  Outliers     OPS  Rounds  Iterations
> -----------------------------------------------------------------------------------------------------------------------------------
> test_benchmark_rispy_big_example_file     16.7548  32.5059  25.0655  6.7390  27.9454  11.2026       2;0  0.0399       5           1
> -----------------------------------------------------------------------------------------------------------------------------------
> 
> Legend:
>   Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
>   OPS: Operations Per Second, computed as 1 / Mean
> ```
> 
> old:
> 
> ```
> ------------------------------------------------------- benchmark: 1 tests -------------------------------------------------------
> Name (time in s)                              Min      Max     Mean  StdDev   Median     IQR  Outliers     OPS  Rounds  Iterations
> ----------------------------------------------------------------------------------------------------------------------------------
> test_benchmark_rispy_big_example_file     30.5232  43.5556  35.9390  5.2012  35.9779  7.8023       2;0  0.0278       5           1
> ----------------------------------------------------------------------------------------------------------------------------------
> 
> Legend:
>   Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
>   OPS: Operations Per Second, computed as 1 / Mean
> ```
> 
> I hope we can speed this up further in the near future (for example, with Rust).

Still a nice speedup, this will be valuable even w/o migrating to 🦀

J535D165 commented:
Thanks for merging. I will continue working on the PubMed parser now. Hopefully, I can implement it without API changes. If that works out, we are ready to prepare v1.
